usr/sbin/pkcsslotd/slotmgr.c

```diff
@@ -113,28 +117,80 @@ int compute_hash(int hash_type, int buf_size, char *buf, char *digest)
     return 0;
 }
 
+/*
+ * This function is called only when running as root user.
+ * Change uid to 'pkcsslotd' and gid to 'pkcs11' and set the group access list
+ * to the groups of user 'pkcsslotd'.
+ * The setuid() call also drops all capabilities (effective and permitted).
+ * The pkcsslotd is then restarted via execv() which applies the capabilities
+ * of the executable file (if any) to the running process, just as if
+ * pkcsslotd was started as pkcsslotd user right away.
+ */
+void drop_privileges(struct passwd *pwd)
+{
+    char program[PATH_MAX + 1] = { 0 };
+    char* args[] = { program, NULL };
+
+    if (readlink("/proc/self/exe", program, PATH_MAX) == -1) {
+        fprintf(stderr, "Failed to get executable file name: %s\n",
+                strerror(errno));
+        exit(1);
+    }
+
+    if (initgroups(pwd->pw_name, pwd->pw_gid) != 0) {
+        fprintf(stderr, "Failed to set the group access list: %s\n",
+                strerror(errno));
+        exit(1);
+    }
+
+    if (setgid(pwd->pw_gid) != 0) {
+        fprintf(stderr, "Failed to set gid to 'pkcs11': %s\n", strerror(errno));
+        exit(1);
+    }
+
+    if (setuid(pwd->pw_uid) != 0) {
+        fprintf(stderr, "Failed to set uid to 'pkcsslotd': %s\n",
+                strerror(errno));
+        exit(1);
+    }
+
+    DbgLog(DL0, "Changed uid from 'root' to 'pkcsslotd' and gid to 'pkcs11'.\n");
+
+    /*
+     * Start pkcsslotd again as pkcsslotd user. This will also apply the
+     * capabilities to those that are set for the executable file (if any).
+     */
+    if (execv(program, args) != 0) {
+        fprintf(stderr, "Failed to re-start pkcsslotd: %s\n",
+                strerror(errno));
+        exit(1);
+    }
+}
+
 /** This function does basic sanity checks to make sure the
  *  eco system is in place for opencryptoki to run properly.
  **/
 void run_sanity_checks(void)
 {
-    int i, ec, uid = -1;
-    struct group *grp = NULL;
+    int i, ec;
+    uid_t uid;
+    gid_t gid;
+    struct passwd *pwd;
+    struct group *grp;
     struct stat sbuf;
     struct dircheckinfo_s dircheck[] = {
-        //drwxrwx---
-        {LOCKDIR_PATH, S_IRWXU | S_IRWXG},
-        {OCK_LOGDIR, S_IRWXU | S_IRWXG},
-        {CONFIG_PATH, S_IRWXU | S_IRWXG},
-        {CONFIG_PATH "/HSM_MK_CHANGE", S_IRWXU | S_IRWXG},
-        {NULL, 0},
+        { "/run/opencryptoki", S_IRWXU | S_IXGRP, 1 },
+        { LOCKDIR_PATH, S_IRWXU | S_IRWXG, 0 },
+        { OCK_LOGDIR, S_IRWXU | S_IRWXG, 0 },
+        { CONFIG_PATH, S_IRWXU | S_IRWXG, 0 },
+        { CONFIG_PATH "/HSM_MK_CHANGE", S_IRWXU | S_IRWXG, 0 },
```

Comment:
As this folder name (/HSM_MK_CHANGE) is global in opencryptoki, we could make a constant for it. Or even a constant in a include file that's included by pkcsslotd and opencryptoki itself.

Reply:
Yea, I should have done that already when introducing the HSM MK change support.
I have added a new commit before the others to change this.